packages/destination-actions/src/destinations/hubspot/upsertCustomObjectRecord/index.ts

```diff
@@ -45,12 +62,61 @@ const action: ActionDefinition<Settings, Payload> = {
     }
   },
   perform: async (request, { payload }) => {
-    return request(`${HUBSPOT_BASE_URL}/crm/v3/objects/${payload.objectType}`, {
-      method: 'POST',
-      json: {
-        properties: flattenObject(payload.properties)
+    // Attempt to search Custom Object record with Custom Search Fields
+    // If Custom Search Fields doesn't have any defined property, skip the search and assume record was not found
+    let searchCustomResponse: ModifiedResponse<SearchResponse> | null = null
+    const hubspotApiClient: Hubspot = new Hubspot(request, payload.objectType)
+    if (
+      typeof payload.customObjectSearchFields === 'object' &&
+      Object.keys(payload.customObjectSearchFields).length > 0
+    ) {
+      try {
+        // Generate custom search payload
+        const responseSortBy: string[] = ['name']
+        searchCustomResponse = await hubspotApiClient.search(
+          { ...payload.customObjectSearchFields },
+          [],
+          responseSortBy
+        )
+      } catch (e) {
+        // HubSpot throws a generic 400 error if an undefined property is used in search
+        // Throw a more informative error instead
+        if ((e as HTTPError)?.response?.status === 400) {
+          throw CustomSearchThrowableError
+        }
+        throw e
       }
-    })
+    }
+    // Store Custom Object Record Id in parent scope
+    // This would be used to store custom object record Id after search or create.
+    let upsertCustomRecordResponse: ModifiedResponse<UpsertRecordResponse>
+    // Check if any custom object record were found based Custom Search Fields
+    // If the search was skipped, searchCustomResponse would have a falsy value (null)
+    if (!searchCustomResponse?.data || searchCustomResponse?.data?.total === 0) {
+      // No existing custom object record found with search criteria, attempt to create a new custom object record
+
+      // setting the createNewCustomRecord default value to true if it's not provided.
+      payload.createNewCustomRecord = !Object.prototype.hasOwnProperty.call(payload, 'createNewCustomRecord')
+        ? true
+        : payload.createNewCustomRecord
+      // If Create New custom object record flag is set to false, skip creation
+      if (!payload.createNewCustomRecord) {
```

Comment:
nit: Don't think its a good idea to update the payload. 
Also [null coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) should do the job here for checking if the field is set if not defaulting to true.
```suggestion
      const createNewCustomRecord = payload?.createNewCustomRecord?? true
      // If Create New custom object record flag is set to false, skip creation
      if (!createNewCustomRecord) {
```